netjsonconfig/backends/zerotier/parser.py

```diff
@@ -0,0 +1,43 @@
+import re
+import tarfile
+from json import loads
+
+from ..base.parser import BaseParser
+
+vpn_pattern = re.compile('^// zerotier controller config:\s', flags=re.MULTILINE)
+config_pattern = re.compile('^([^\s]*) ?(.*)$')
+config_suffix = '.json'
+
+
+class ZeroTierParser(BaseParser):
+    def parse_text(self, config):
+        return {'zerotier': self._get_vpn_config(config)}
+
+    def parse_tar(self, tar):
+        fileobj = tar.buffer if hasattr(tar, 'buffer') else tar
+        tar = tarfile.open(fileobj=fileobj)
+        text = ''
+        for member in tar.getmembers():
+            if not member.name.endswith(config_suffix):
+                continue
+            text += '// zerotier controller config: {name}\n\n{contents}\n'.format(
+                **{
+                    'name': member.name,
+                    'contents': tar.extractfile(member).read().decode(),
+                }
+            )
+        return self.parse_text(text)
+
+    def _get_vpn_config(self, text):
+        vpn_configs = []
+        vpn_instances = vpn_pattern.split(text)
+        if '' in vpn_instances:
+            vpn_instances.remove('')
+        for vpn in vpn_instances:
+            lines = vpn.split('\n')
+            text_contents = '\n'.join(lines[2:])
+            # do not end with double new line
+            if text_contents.endswith('\n\n'):
+                text_contents = text_contents[0:-1]
+            vpn_configs.append(loads(text_contents))
+        return vpn_configs
```

Comment:
Can we use `json.loads` directly here? What are the drawbacks of doing that?

Reply:
I have used [`json.loads` method](https://github.com/openwisp/netjsonconfig/pull/272/files#diff-4e782cb854b8f20ae055ed735e67603137d6014ed8a1f73f8ba00b7ca107a1baR39) after extracting a valid JSON string (i.e., after removing `vpn_pattern` and separating multiple zerotier text configurations). Otherwise, if we directly use `json.loads` without filtering, it will throw an error `json.decoder.JSONDecodeError: Extra data:` and parsing will fail.